[c10d] disable compute_duration by default

shuqiangzhang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #122138

Summary:
Compute duration would invoke additional cuda overhead and possibly
GPU mem increase and possible hang, so we want to disable it by default and enable it only
when needed, or at least when timing is enabled.

Test Plan:
Test with existing unit test

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang

shuqiangzhang committed:
[c10d] disable compute_duration by default
Summary:
Compute duration would invoke additional cuda overhead and possibly
hang, so we want to disable it by default and enable it only
when needed, or at least when timing is enabled.

Test Plan:
Test with existing unit test

[ghstack-poisoned]

wconstab commented:
not sure if your change does anything meaningful?

iirc TIMING_ENABLED env is disabled by default, so it should be disabled during all your testing.  (1. was it enabled?)

then if TIMING_ENABLED is false, every work obj should be missing its start events.  The missing start events should cause the compute_duration to be skipped automatically.  (it would be basically like a nullptr error if we tried to compute duration while timing was disabled, and start events didn't exist).

So i'm not too surprised if computeDuration is implicated in this peak-memory mess since I already suspected some kind of cuda event issue.  But at the same time, it'd be nice to understand more carefully what exactly is going on.  Is the test case actually adversarial in some way, or why does calling computeDuration lead to such a peak memory increase?  Naively, i'd only expect computeDuration to add some runtime overhead.

shuqiangzhang committed:
Update on "[c10d] disable compute_duration by default"


Summary:
Compute duration would invoke additional cuda overhead and possibly
GPU mem increase and possible hang, so we want to disable it by default and enable it only
when needed, or at least when timing is enabled.

Test Plan:
Test with existing unit test

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang

[ghstack-poisoned]

shuqiangzhang commented:
Verified that TIMING_ENABLED is true in the tests, startEvent and endEvent is NOT null and getDurationFromEvent() did get called. The mem increase with duration calculation can be reproed consistently. Hence, I believe we should disable duration calculation for now and enable it when it is really needed.

shuqiangzhang committed:
Update on "[c10d] disable compute_duration by default"


Summary:
Compute duration would invoke additional cuda overhead and possibly
GPU mem increase and possible hang, so we want to disable it by default and enable it only
when needed, or at least when timing is enabled.

Test Plan:
Test with existing unit test

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang

[ghstack-poisoned]

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -746,6 +746,7 @@ ProcessGroupNCCL::ProcessGroupNCCL(
       getCvarInt(TORCH_NCCL_WAIT_TIMEOUT_DUMP_MILSEC, 60 * 1000 /*60 Sec*/);
   coordCheckIntervalMilSec_ = getCvarInt(TORCH_NCCL_COORD_CHECK_MILSEC, 1000);
   ncclTraceBufferSize_ = getCvarInt(TORCH_NCCL_TRACE_BUFFER_SIZE, 0);
+  computeDuratoin_ = getCvarBool(TORCH_NCCL_COMPUTE_DURATION, false);
```
typo.

shuqiangzhang replied:
my bad

wconstab commented on torch/csrc/distributed/c10d/TraceUtils.h:
```diff
@@ -528,13 +528,11 @@ struct NCCLTraceBuffer {
-  but care should be taken to avoid computing duration in any function that must
-  never hang. (timing must also be enabled for compute_duration - see
-  TORCH_NCCL_ENABLE_TIMING).
+  compute_duration defaults to false since calling cuda APIs may hang,
```
maybe also mention the peak memory issue?

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -811,6 +812,7 @@ ProcessGroupNCCL::ProcessGroupNCCL(
             << waitTimeoutDumpInMilSec_
             << ", TORCH_NCCL_DESYNC_DEBUG: " << desyncDebug_
             << ", TORCH_NCCL_ENABLE_TIMING: " << enableTiming_.load()
+            << ", TORCH_NCCL_COMPUTE_DURATION: " << computeDuratoin_
```
i dont see any logic to error if someone sets compute_duration=True and timing_enabled=False.  That should be an error, right?  (or else it should be documented that setting 'compute_duration' automatically enables timing.

shuqiangzhang replied:
Yes, decided to go the later option to make user's life easier such that they just need t specify one env to enable compute_duration

wconstab reviewed (changes requested):
at least fix the var name, then i think its landable. hope we can delete this env soon. it's getting crazy!

shuqiangzhang committed:
Update on "[c10d] disable compute_duration by default"


Summary:
Compute duration would invoke additional cuda overhead and possibly
GPU mem increase and possible hang, so we want to disable it by default and enable it only
when needed, or at least when timing is enabled.

Test Plan:
Test with existing unit test

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang

[ghstack-poisoned]

shuqiangzhang commented:
@pytorchbot merge

shuqiangzhang committed:
Update on "[c10d] disable compute_duration by default"


Summary:
Compute duration would invoke additional cuda overhead and possibly
GPU mem increase and possible hang, so we want to disable it by default and enable it only
when needed, or at least when timing is enabled.

Test Plan:
Test with existing unit test

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang

[ghstack-poisoned]

shuqiangzhang commented:
@pytorchbot merge
